src/python/pim/polynomial_representation.py

```diff
@@ -0,0 +1,89 @@
+from typing import List
+import re
+import itertools as it
+import numpy as np
+
+class Polynomial:
+    def __init__(self, coefficients: List[int]) -> None:
+        self.coefficients: List[int] = coefficients
+
+    def __str__(self) -> str:
+        """returns a string representation of the polynomial"""
+        coeff_list: List[str] = [display(c, i) for i, c in enumerate(self.coefficients)]
+        filter_coeff_list = list(filter(lambda x: x != "", coeff_list))
+        return re.sub(r"\+ -", "- ", " + ".join(filter_coeff_list))
+
+    def __add__(self, other: "Polynomial") -> "Polynomial":
+        """returns the sum of two polynomials (like degrees are added together)"""
+        return [sum(n) for n in it.zip_longest(self.coefficients, other.coefficients, fillvalue = 0)]
+
+    def __mul__(self, other: "Polynomial") -> "Polynomial":
+        """returns the product of two polynomials"""
+        len_self: int = len(self.coefficients)
+        len_other: int = len(other.coefficients)
+        blank_coeff_list: list(int) = [0] * (len_self + len_other - 1)
```

Comment:
Do we know that the resulting `Polynomial` will always include `len_self + len_other - 1` terms after simplification? Feels like a nice opportunity for a proof.

Reply:
So we had actually already proved a few things when we did the [chapter 2.1 exercise](https://github.com/kserapea/pim/blob/feature/secret_sharing/Chapter%202%20Exercises.ipynb).

- If 𝑓 is a degree 2 polynomial and 𝑔 is a degree 1 polynomial then their product 𝑓⋅𝑔 is a degree 3 polynomial
- If 𝑓 is a degree 𝑛 polynomial and 𝑔 is a degree 𝑚 polynomial then their product 𝑓⋅𝑔 is a degree 𝑛+𝑚 polynomial.
- If 𝑓 or 𝑔 are a zero polynomial, is their product 𝑓⋅𝑔 is a degree 𝑚 polynomial (convention that a zero polynomial has a degree of 0).

I will think through the code to ensure it actually covers those cases instead of using my unproven (and subsequently disproved method).

Reply:
I realize now that what my 
`len_self` + `len_other` - 1 
was accounting for was that inherently, when you multiply x^1 * x^1, you end up with x^2, (which is of length 1); therefore I was automatically discarding the "0x" that would be present in a coefficient list. So, my thinking is not wrong for polynomials in general, but if we are measuring the length of a coefficient list and including 0's as placeholders to account for degrees, than the length aspect that way does not work. (you probably saw that right away, but I was thinking from the maths side and not from the data structures side...)

Reply:
I don't think I saw it clearly right away, just had a nagging feeling. Thanks for confirming!